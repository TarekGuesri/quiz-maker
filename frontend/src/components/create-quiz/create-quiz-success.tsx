import React, { FC } from "react";

export const CreateQuizSuccess: FC = () => {
  return <div>create-quiz-success</div>;
};
