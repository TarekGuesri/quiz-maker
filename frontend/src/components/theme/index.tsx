import React, { FC } from 'react';
import CssBaseline from '@mui/material/CssBaseline';
import PropTypes from 'prop-types';
import { createTheme, ThemeProvider } from '@mui/material/styles';
import { useAppSelector } from 'src/redux/hooks';
import varelaRound from 'src/assets/fonts/VarelaRound-Regular.ttf';

export const Theme: FC = (props) => {
  const { darkMode } = useAppSelector((state) => state.ui);

  const theme = createTheme({
    palette: {
      mode: darkMode ? 'dark' : 'light',
      background: {
        default: darkMode ? '#333' : '#f7f7f7',
        paper: darkMode ? '#333' : '#fbfbfb',
      },
      primary: {
        main: '#8e49de',
        // contrastText:
      },
      secondary: {
        main: '#f9de4a',
      },
      contrastThreshold: 3,
      tonalOffset: 0.2,
    },
    typography: {
      fontFamily: 'Varela Round',
    },
    components: {
      MuiCssBaseline: {
        styleOverrides: `
          @font-face {
            font-family: 'Varela Round';
            font-style: normal;
            font-display: swap;
            font-weight: 400;
            src: url(${varelaRound}) format('truetype');
        `,
      },
    },
  });

  return (
    <ThemeProvider theme={theme}>
      {props.children}
      <CssBaseline />
    </ThemeProvider>
  );
};

Theme.propTypes = {
  children: PropTypes.node.isRequired,
};
