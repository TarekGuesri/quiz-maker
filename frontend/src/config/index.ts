import { Environment } from "../types";

// Environment getter
export const getEnv: () => Environment = () => {
  switch (location.hostname) {
    case "localhost":
      return "development";
    default:
      return "production";
  }
};

const frontendPort = location.port;
const environments: Environment[] = ["development", "production"];
const e = environments.indexOf(getEnv());

export const config = {
  frontend: {
    url: [
      `http://localhost:${frontendPort}`, // Development
      "https://quiz-maker.tariqguesri.com", // Production
    ][e],
  },
  api: {
    url: [
      "http://localhost:5000", // Development
      "https://quiz-maker.tariqguesri.com", // Production
    ][e],
  },
};
