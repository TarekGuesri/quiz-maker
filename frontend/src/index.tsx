import React from "react";
import ReactDOM from "react-dom";
import { Router } from "react-router-dom";
import { history } from "src/utils/history";
import { Provider } from "react-redux";

import { store } from "./redux/store";
import App from "./App";

ReactDOM.render(
  <React.StrictMode>
    <Provider store={store}>
      <Router history={history}>
        <App />
      </Router>
    </Provider>
  </React.StrictMode>,
  document.getElementById("root"),
);
