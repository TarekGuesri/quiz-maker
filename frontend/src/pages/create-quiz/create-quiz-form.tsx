import React from "react";
import { useAppDispatch, useAppSelector } from "src/redux/hooks";
import { createStyles, makeStyles } from "@mui/styles";
import { Theme } from "@mui/material";
import Alert from "@mui/material/Alert";
import Divider from "@mui/material/Divider";
import Grid from "@mui/material/Grid";
import TextField from "@mui/material/TextField";
import Typography from "@mui/material/Typography";
import Radio from "@mui/material/Radio";
import RadioGroup from "@mui/material/RadioGroup";
import FormControlLabel from "@mui/material/FormControlLabel";
import FormControl from "@mui/material/FormControl";
import FormLabel from "@mui/material/FormLabel";
import Pagination from "@mui/material/Pagination";
import Button from "@mui/material/Button";
import Tooltip from "@mui/material/Tooltip";
import {
  addQuestion,
  changeAnswer,
  changeTitle,
  changePage,
  changeQuestion,
  createQuiz,
  removeQuestion,
  setSelectedAnswer,
} from "src/redux/create-quiz/create-quiz-slice";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    formLabel: {
      color: theme.palette.text.primary,
      fontSize: "1.2rem",
      "&.Mui-focused": {
        color: theme.palette.text.primary,
      },
    },
    helperText: {
      color: theme.palette.text.secondary,
      marginBottom: theme.spacing(2),
    },
    formControlLabel: {
      marginBottom: theme.spacing(2),
    },
    pagination: {
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
    },
    button: {
      width: "190px",
      padding: "10px",
    },
    buttonGridItem: {
      paddingLeft: "0!important",
    },
  }),
);

export const CreateQuizForm: React.FC = ({}) => {
  const classes = useStyles();
  const dispatch = useAppDispatch();
  const {
    title,
    questionIndex,
    questions,
    selectedAnswers,
    isLoading,
    errorMessage,
  } = useAppSelector((state) => state.createQuiz);

  const handleSelectAnswer = (event: React.ChangeEvent<HTMLInputElement>) => {
    const selectedAnswer = (event.target as HTMLInputElement).value;

    dispatch(setSelectedAnswer(selectedAnswer));
  };

  const handleChangePage = (
    _event: React.ChangeEvent<unknown>,
    value: number,
  ) => {
    if (value) {
      dispatch(changePage(value - 1));
    }
  };

  const handleChangeTitle = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;

    dispatch(changeTitle(value));
  };

  const handleChangeQuestion = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;

    dispatch(changeQuestion(value));
  };

  const handleChangeAnswer = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { value, name } = event.target;

    dispatch(changeAnswer({ name, value }));
  };

  const currentQuestion = questions[questionIndex];

  // We use this for selected answer's radio button
  const selectedAnswer = selectedAnswers[questionIndex];

  return (
    <>
      <Typography variant="h5" component="div" mb={4}>
        Create a quiz
      </Typography>

      <Grid container>
        {/* Quiz Title */}
        <Grid item xs={12} mb={3}>
          <TextField
            value={title}
            onChange={handleChangeTitle}
            label="Quiz Title"
            variant="outlined"
            size="medium"
            fullWidth
          />
        </Grid>

        {/* Quiz Title */}
        <Grid item xs={12} mb={3}>
          <Divider />
        </Grid>

        {/* Question */}
        <Grid item xs={12} mb={3}>
          <TextField
            value={currentQuestion.content}
            onChange={handleChangeQuestion}
            label={`Question ${questionIndex + 1}`}
            variant="outlined"
            size="medium"
            fullWidth
          />
        </Grid>

        {/* Answers */}
        <Grid item xs={12} mb={2}>
          <FormControl component="fieldset">
            <FormLabel className={classes.formLabel}>
              Answers
              <Typography className={classes.helperText}>
                Check the correct answer
              </Typography>
            </FormLabel>

            <RadioGroup
              aria-label="gender"
              name="controlled-radio-buttons-group"
              value={selectedAnswer}
              onChange={handleSelectAnswer}
            >
              {currentQuestion.answers.map((answer, index) => (
                <FormControlLabel
                  key={answer.id}
                  className={classes.formControlLabel}
                  value={answer.id}
                  control={<Radio />}
                  label={
                    <TextField
                      label={`Answer ${index + 1}...`}
                      name={answer.id}
                      value={answer.text}
                      onChange={handleChangeAnswer}
                    />
                  }
                />
              ))}
            </RadioGroup>
          </FormControl>
        </Grid>

        {/* Pagination */}
        <Grid item xs={12} mb={4}>
          <Pagination
            className={classes.pagination}
            count={questions.length}
            page={questionIndex + 1}
            color="primary"
            onChange={handleChangePage}
          />
        </Grid>

        {/* Error Message */}
        {errorMessage && (
          <Grid item xs={12} mb={4}>
            <Alert variant="filled" severity="error">
              {errorMessage}
            </Alert>
          </Grid>
        )}
        {/* Buttons */}
        <Grid
          container
          item
          direction="row"
          xs={12}
          maxWidth={700}
          mx="auto"
          spacing={3}
        >
          <Grid item xs={12} md={4} className={classes.buttonGridItem}>
            <Button
              variant="contained"
              color="primary"
              className={classes.button}
              onClick={() => dispatch(removeQuestion())}
              disabled={questions.length < 2}
            >
              Remove Question
            </Button>
          </Grid>
          <Grid item xs={12} md={4} className={classes.buttonGridItem}>
            <Tooltip title="You can only have 10 questions max per quiz">
              <div>
                <Button
                  variant="contained"
                  color="primary"
                  className={classes.button}
                  onClick={() => dispatch(addQuestion())}
                  disabled={questions.length >= 10}
                >
                  Add a question
                </Button>
              </div>
            </Tooltip>
          </Grid>
          <Grid item xs={12} md={4} className={classes.buttonGridItem}>
            <Button
              variant="contained"
              color="secondary"
              className={classes.button}
              onClick={() => dispatch(createQuiz())}
              disabled={isLoading}
            >
              Create Quiz
            </Button>
          </Grid>
        </Grid>
      </Grid>
    </>
  );
};
