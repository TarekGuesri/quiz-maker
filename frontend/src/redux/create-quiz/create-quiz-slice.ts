import axios, { AxiosError, AxiosResponse } from "axios";
import { createSlice, PayloadAction, current } from "@reduxjs/toolkit";
import { v4 as uuidv4 } from "uuid";
import { AppThunk, RootState } from "src/redux/store";
import { CreateQuizState, ChangeAnswer } from "src/types";

const firstAnswerID = uuidv4();

const initialState: CreateQuizState = {
  title: "aaaa",
  questionIndex: 0,
  questions: [
    {
      id: uuidv4(),
      content: "bbbb",
      answers: [
        { id: firstAnswerID, text: "1111" },
        { id: uuidv4(), text: "222" },
        { id: uuidv4(), text: "333" },
        { id: uuidv4(), text: "4444" },
      ],
    },
  ],
  selectedAnswers: [firstAnswerID],
  isLoading: false,
  isValid: false,
  errorMessage: "",
};

export const createQuizSlice = createSlice({
  name: "create-quiz",
  initialState,
  reducers: {
    addQuestion: (state) => {
      const firstAnswerID = uuidv4();

      // Adding the question
      state.questions.push({
        id: uuidv4(),
        content: "",
        answers: [
          { id: firstAnswerID, text: "" },
          { id: uuidv4(), text: "" },
          { id: uuidv4(), text: "" },
          { id: uuidv4(), text: "" },
        ],
      });

      // Adding new selected answer
      state.selectedAnswers.push(firstAnswerID);

      // Changing index
      state.questionIndex = current(state.questions).length - 1;
    },
    removeQuestion: (state) => {
      // We remove the selected answer from the target question
      const currentIndex = state.questionIndex;
      state.selectedAnswers.splice(currentIndex, 1);

      // We remove the question
      state.questions.splice(currentIndex, 1);

      // We change the index
      const newQuestionsLength = current(state.questions).length;

      state.questionIndex =
        currentIndex === newQuestionsLength ? currentIndex - 1 : currentIndex;
    },
    changeTitle: (state, action: PayloadAction<string>) => {
      state.title = action.payload;
    },
    changeQuestion: (state, action: PayloadAction<string>) => {
      state.questions[state.questionIndex].content = action.payload;
    },
    changeAnswer: (state, action: PayloadAction<ChangeAnswer>) => {
      const { value, name } = action.payload;

      // We get the answer's index
      const answerIndex = state.questions[
        state.questionIndex
      ].answers.findIndex((answer) => answer.id === name);

      // We change the answer's text
      state.questions[state.questionIndex].answers[answerIndex].text = value;
    },
    changePage: (state, action: PayloadAction<number>) => {
      state.questionIndex = action.payload;
    },
    setSelectedAnswer: (state, action: PayloadAction<string>) => {
      state.selectedAnswers[state.questionIndex] = action.payload;
    },
    validateForm: (state) => {
      // We clear errors
      state.errorMessage = "";

      // Then we validate the form
      const { title, questions } = current(state);

      // Validating the title
      if (!title) {
        state.errorMessage = "Quiz title can't be empty";
        state.isValid = false;
        return;
      }

      // Validating questions
      const nonValidQuestions: number[] = [];

      questions.forEach((question, index) => {
        const { content, answers } = question;
        // We validate the question's content
        if (!content) {
          nonValidQuestions.push(index + 1);
          return;
        }

        // Validating question's answers
        for (let i = 0; i < answers.length; i++) {
          const answer = answers[i];

          if (!answer.text) {
            nonValidQuestions.push(index + 1);
            return;
          }
        }
      });

      if (nonValidQuestions.length > 0) {
        state.errorMessage = `Please fill in all the fields of the questions "${nonValidQuestions.join(
          ", ",
        )}"`;
        state.isValid = false;
        return;
      } else {
        state.isValid = true;
      }
    },
    setLoading: (state) => {
      state.isLoading = true;
    },
    createQuizSuccess: (state, action: PayloadAction<string>) => {
      console.log(action.payload);

      state.errorMessage = "";
      state.isLoading = false;
    },
    createQuizFail: (state, action: PayloadAction<string>) => {
      state.errorMessage = action.payload;
      state.isLoading = false;
    },
  },
});

export const {
  addQuestion,
  changeAnswer,
  changeTitle,
  changePage,
  changeQuestion,
  validateForm,
  removeQuestion,
  setSelectedAnswer,
  setLoading,
  createQuizSuccess,
  createQuizFail,
} = createQuizSlice.actions;

export const selectCeateQuiz = (state: RootState) => state.createQuiz;

export const createQuiz = (): AppThunk => async (dispatch, getState) => {
  dispatch(validateForm());

  const { title, questions, selectedAnswers, isValid } = selectCeateQuiz(
    getState(),
  );

  // If the form is valid, we send a request to the api
  if (isValid) {
    dispatch(setLoading());

    try {
      const res: AxiosResponse = await axios.post("quizzes", {
        title,
        questions,
        selectedAnswers,
      });

      dispatch(createQuizSuccess(res.data));
    } catch (error) {
      // Catching the error
      const { response } = error as AxiosError;

      dispatch(
        createQuizFail(response?.data || "Something unexpected happend!"),
      );
    }
  }
};

export default createQuizSlice.reducer;
