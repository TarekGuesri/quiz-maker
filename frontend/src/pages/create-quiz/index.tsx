import React from "react";
import { useAppSelector } from "src/redux/hooks";
import { CardBox } from "src/components/cards/card-box";
import { CreateQuizForm } from "src/components/create-quiz/create-quiz-form";
import { CreateQuizSuccess } from "src/components/create-quiz/create-quiz-success";

const CreateQuiz: React.FC = ({}) => {
  const { quizID } = useAppSelector((state) => state.createQuiz);

  return (
    <CardBox>{quizID ? <CreateQuizSuccess /> : <CreateQuizForm />}</CardBox>
  );
};

export default CreateQuiz;
