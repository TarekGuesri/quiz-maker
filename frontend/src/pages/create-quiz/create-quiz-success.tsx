import React, { FC } from "react";
import { useAppDispatch } from "src/redux/hooks";
import Typography from "@mui/material/Typography";
import CheckCircleIcon from "@mui/icons-material/CheckCircle";
import Button from "@mui/material/Button";
import Grid from "@mui/material/Grid";
import Link from "@mui/material/Link";
import { resetState } from "src/redux/create-quiz/create-quiz-slice";
import { config } from "src/config";

interface iProps {
  quizID: string;
}

export const CreateQuizSuccess: FC<iProps> = ({ quizID }) => {
  const dispatch = useAppDispatch();
  const quizURL = `${config.frontend.url}/quiz/${quizID}`;

  return (
    <>
      <Typography variant="h5" component="div" mb={1}>
        Your quiz has been created!
      </Typography>

      <Typography color="primary" mb={4}>
        <CheckCircleIcon fontSize="large" />
      </Typography>

      <Grid container>
        {/* Quiz URL */}
        <Grid item xs={12} mb={5}>
          Here&apos;s the link to your quiz:{" "}
          <Link href={quizURL}>{quizURL}</Link>
        </Grid>

        {/* Create another quiz button */}
        <Grid item xs={12} mb={1}>
          <Button
            color="primary"
            variant="contained"
            size="large"
            onClick={() => dispatch(resetState())}
          >
            Create another quiz
          </Button>
        </Grid>
      </Grid>
    </>
  );
};
