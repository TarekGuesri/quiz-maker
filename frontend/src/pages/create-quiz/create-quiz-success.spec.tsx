/**
 * @jest-environment jsdom
 */

import React from "react";
import { cleanup, render, screen, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import { Theme } from "src/components/theme";
import { store } from "src/redux/store";
import { config } from "src/config";

import { CreateQuizSuccess } from "./create-quiz-success";
import { createQuizSuccess } from "src/redux/create-quiz/create-quiz-slice";

describe("src/components/create-quiz-success.tsx", () => {
  afterEach(() => cleanup());

  test("Render CreateQuizSuccess component", () => {
    const renderedComponent = render(
      <Provider store={store}>
        <Theme>
          <CreateQuizSuccess quizID="test" />
        </Theme>
      </Provider>,
    );

    expect(renderedComponent.container).toMatchSnapshot();
  });

  test("The quiz url must have the same quizID from props", () => {
    const quizID = "test";

    render(
      <Provider store={store}>
        <Theme>
          <CreateQuizSuccess quizID={quizID} />
        </Theme>
      </Provider>,
    );

    const expectedQuizURL = `${config.frontend.url}/quiz/${quizID}`;

    screen.getByRole("link", {
      name: expectedQuizURL,
    });
  });

  test("quizID from redux becomes empty when clicking create another quiz button", () => {
    store.dispatch(createQuizSuccess("test"));

    const {
      createQuiz: { quizID },
    } = store.getState();

    expect(quizID).toEqual("test");

    render(
      <Provider store={store}>
        <Theme>
          <CreateQuizSuccess quizID={quizID} />
        </Theme>
      </Provider>,
    );

    const createButton = screen.getByRole("button", {
      name: /create another quiz/i,
    });

    fireEvent.click(createButton);

    expect(store.getState().createQuiz.quizID).toEqual("");
  });
});
