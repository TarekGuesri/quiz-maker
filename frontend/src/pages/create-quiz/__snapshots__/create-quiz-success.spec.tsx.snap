// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/create-quiz-success.tsx Render CreateQuizSuccess component 1`] = `
<div>
  <div
    class="MuiTypography-root MuiTypography-h5 css-ue653o-MuiTypography-root"
  >
    Your quiz has been created!
  </div>
  <p
    class="MuiTypography-root MuiTypography-body1 css-eshe5m-MuiTypography-root"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
      data-testid="CheckCircleIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
      />
    </svg>
  </p>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-11lq3yg-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-77ccod-MuiGrid-root"
    >
      Here's the link to your quiz:
       
      <a
        class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-121d6ld-MuiTypography-root-MuiLink-root"
        href="http://localhost:/quiz/test"
      >
        http://localhost:/quiz/test
      </a>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1kgfixx-MuiGrid-root"
    >
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root css-17gny3x-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Create another quiz
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
